extensions.bzl

```diff
@@ -0,0 +1,71 @@
+load("//:rules.bzl", bb_macro = "buildbuddy")
+
+def _ext_impl(mctx):
+    found_root_module = False
+    root_module = struct()
+    for m in mctx.modules:
+        if m.is_root:
+            found_root_module = True
+            root_module = m
+            break
+    if not found_root_module:
+        fail("buildbuddy must be used in root module")
+
+    macro_args = dict()
+    if len(root_module.tags.platform) > 1:
+        fail("buildbuddy.platform can only be specified once per module")
+    elif len(root_module.tags.platform) == 1:
+        platform_tag = root_module.platform[0]
+        macro_args |= {
+            "container_image": platform_tag.container_image,
+        }
+    if len(root_module.tags.gcc_toolchain) > 1:
+        fail("buildbuddy.gcc_toolchain can only be specified once per module")
+    elif len(root_module.tags.gcc_toolchain) == 1:
+        gcc_toolchain_tag = root_module.gcc_toolchain[0]
+        macro_args |= {
+            "gcc_major_version": gcc_toolchain_tag.gcc_major_version,
+            "extra_cxx_builtin_include_directories": gcc_toolchain_tag.extra_cxx_builtin_include_directories,
+        }
+    if len(root_module.tags.msvc_toolchain) > 1:
+        fail("buildbuddy.msvc_toolchain can only be specified once per module")
+    elif len(root_module.tags.msvc_toolchain) == 1:
+        msvc_toolchain_tag = root_module.msvc_toolchain[0]
+        macro_args |= {
+            "msvc_edition": msvc_toolchain_tag.msvc_edition,
+            "msvc_release": msvc_toolchain_tag.msvc_release,
+            "msvc_version": msvc_toolchain_tag.msvc_version,
+            "windows_kits_release": msvc_toolchain_tag.windows_kits_release,
+            "windows_kits_version": msvc_toolchain_tag.windows_kits_version,
+        }
```

Comment:
Sorry, I imagined this differently and `elif` wasn't the right lead. This is the separation of validation and "business logic" I wanted to get to, let me know what you think:

```suggestion
    tags = root_module.tags
    if len(tags.platform) > 1:
        fail("buildbuddy.platform can only be specified once per module")
    platform_tag = tags.platform[0] if tags.platform else None
    if len(tags.gcc_toolchain) > 1:
        fail("buildbuddy.gcc_toolchain can only be specified once per module")
    gcc_toolchain_tag = tags.gcc_toolchain[0] if tags.gcc_toolchain else None
    if len(tags.msvc_toolchain) > 1:
        fail("buildbuddy.msvc_toolchain can only be specified once per module")
    msvc_toolchain_tag = tags.msvc_toolchain[0] if tags.msvc_toolchain else None

    macro_args = dict()
    if platform_tag:
        macro_args |= {
            "container_image": platform_tag.container_image,
        }
    if gcc_toolchain_tag:
        macro_args |= {
            "gcc_major_version": gcc_toolchain_tag.gcc_major_version,
            "extra_cxx_builtin_include_directories": gcc_toolchain_tag.extra_cxx_builtin_include_directories,
        }
    if msvc_toolchain_tag:
        macro_args |= {
            "msvc_edition": msvc_toolchain_tag.msvc_edition,
            "msvc_release": msvc_toolchain_tag.msvc_release,
            "msvc_version": msvc_toolchain_tag.msvc_version,
            "windows_kits_release": msvc_toolchain_tag.windows_kits_release,
            "windows_kits_version": msvc_toolchain_tag.windows_kits_version,
        }
```

Reply:
a bit more "if" but I don't hate it.